src/system/vscode/asyncDebouncer.ts

```diff
@@ -0,0 +1,134 @@
+import type { CancellationToken, Disposable } from 'vscode';
+import { CancellationTokenSource } from 'vscode';
+import { CancellationError } from '../../errors';
+import type { Deferrable } from '../function';
+import type { Deferred } from '../promise';
+import { defer } from '../promise';
+
+export interface AsyncTask<T> {
+	(cancelationToken: CancellationToken): T | Promise<T>;
+}
+
+/**
+ * This is similar to `src/system/function.ts: debounce` but it's for async tasks.
+ * The old `debounce` function does not awaits for promises, so it's not suitable for async tasks.
+ *
+ * Here the debouncer returns a promise that awaits task for completion.
+ * Also we can let tasks know if they are cancelled by passing a cancellation token.
+ *
+ * Despite being able to accept synchronous tasks, we always return a promise here. It's implemeted this way for simplicity.
+ */
+export function createAsyncDebouncer<T>(delay: number): Disposable & Deferrable<(task: AsyncTask<T>) => Promise<T>> {
```

Comment:
Would it work to move this to the same `src/system/function` file? It appears to me like this is another utility for working with functions.

Reply:
@axosoft-ramint 

I wanted to do this. But `src/system/function.ts` is supposed to be used both in vscode and in web-view. While `asyncDebouncer` imports `CancellationTokenSource` from `vscode`, so placing it to `function.ts` makes the project non-compilable.

Therefore I've put it under `src/system/vscode`.

Reply:
I see. Thanks for explaining 👍🏼